labs/lab05/src/UI.java

```diff
@@ -0,0 +1,136 @@
+import java.util.*;
+
+public class UI {
+    private Grammar grammar = null;
+    private boolean running = true;
+    private final Map<String, String> options = new LinkedHashMap<>();
+    private final Map<String, Runnable> optionsHandlers = new LinkedHashMap<>(); // may contain null handlers
+
+    public UI() {
+        addOptions();
+    }
+
+    //region running & handling options
+    public void run() {
+        Scanner scanner = new Scanner(System.in);
+
+        while (running) {
+            options.forEach((option, description) -> System.out.println(option + ". " + description));
+            System.out.print("\t Your option: ");
+
+            handleUserOption(scanner.nextLine());
+            System.out.println();
+        }
+    }
+
+    /**
+     * @param option the option given by the user
+     */
+    private void handleUserOption(String option) {
+        if (Objects.equals(option, "0")) {
+            System.out.println("Exiting...");
+            running = false;
+        } else if (options.containsKey(option)) {
+            chooseAndCallOptionHandler(option);
+        } else {
+            System.out.println("Invalid option");
+        }
+    }
+
+    private void chooseAndCallOptionHandler(String option) {
+        if (grammar == null && !Objects.equals(option, "1")) {
+            System.out.println("You must read the grammar first!");
```

Comment:
the option could be showed to the user only after a grammar is chosen, eliminating the need for this check

Reply:
good remark.
one approach would be: at the start of the while loop, before printing the menu, if the grammar is null, add the option, otherwise remove it